segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -538,6 +546,25 @@ void notifyIndexOffsetChanged(long segmentId, long indexOffset) {
                 ignored -> CompletableFuture.completedFuture(this.cache.getTailHashes(segment.getSegmentId())));
     }
 
+    /**
+     * Gets the {@link SegmentSortedKeyIndex} associated with the given Segment, as provided by
+     * {@link ContainerSortedKeyIndex#getSortedKeyIndex}. This can be used to safely iterate through Keys of a Sorted
+     * Table Segment while including both fully indexed and tail updates.
+     *
+     * Note: this will return the same result as {@link ContainerSortedKeyIndex#getSortedKeyIndex}, however this method
+     * will wait on any Segment-specific recovery (and trigger it) to complete before executing, which should enable a
+     * safe iteration for recently recovered Table Segments.
+     *
+     * @param segment A {@link DirectSegmentAccess} representing the Segment for which to get the {@link SegmentSortedKeyIndex}.
+     * @return A CompletableFuture that, when completed, will contain the desired result. This Future will wait on any
```

Comment:
> This Future will wait on any Segment-specific recovery to complete before executing.

This is a good information to have, but it made me wonder if that's not true of any segment-related operation. Is there anything special about this call to get the sorted key index with respect to recovery?

Reply:
It's true of any segment-related operation that have this phrase attached to them. Not all operations need to wait on that. Conditional updates must wait, but unconditional updates do not. Retrievals and iterators must wait.